src/responder/kcm/kcmsrv_ccache_key.c

```diff
@@ -0,0 +1,145 @@
+
+#include "util/util.h"
+#include "responder/kcm/kcmsrv_ccache_pvt.h"
+
+/*
+ * The secrets store is a key-value store at heart. We store the UUID
+ * and the name in the key to allow easy lookups be either key
+ */
+#define SEC_KEY_SEPARATOR   '-'
+
+const char *sec_key_create(TALLOC_CTX *mem_ctx,
+                           const char *name,
+                           uuid_t uuid)
+{
+    char uuid_str[UUID_STR_SIZE];
+
+    uuid_unparse(uuid, uuid_str);
+    return talloc_asprintf(mem_ctx,
+                           "%s%c%s", uuid_str, SEC_KEY_SEPARATOR, name);
+}
+
+static bool sec_key_valid(const char *sec_key)
+{
+    if (sec_key == NULL) {
+        return false;
+    }
+
+    if (strlen(sec_key) < UUID_STR_SIZE + 1) {
+        /* One char for separator (at UUID_STR_SIZE, because strlen doesn't
+         * include the '\0', but UUID_STR_SIZE does) and at least one for
+         * the name */
+        DEBUG(SSSDBG_CRIT_FAILURE, "Key %s is too short\n", sec_key);
+        return false;
+    }
+
+    if (sec_key[UUID_STR_SIZE - 1] != SEC_KEY_SEPARATOR) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Key doesn't contain the separator\n");
+        return false;
+    }
+
+    return true;
+}
+
+errno_t sec_key_parse(TALLOC_CTX *mem_ctx,
+                      const char *sec_key,
+                      const char **_name,
+                      uuid_t uuid)
+{
+    char uuid_str[UUID_STR_SIZE];
+
+    if (!sec_key_valid(sec_key)) {
+        return EINVAL;
+    }
+
+    strncpy(uuid_str, sec_key, sizeof(uuid_str)-1);
+    if (sec_key[UUID_STR_SIZE - 1] != SEC_KEY_SEPARATOR) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Key doesn't contain the separator\n");
+        return EINVAL;
+    }
+    uuid_str[UUID_STR_SIZE-1] = '\0';
+
+    *_name = talloc_strdup(mem_ctx, sec_key + UUID_STR_SIZE);
+    if (*_name == NULL) {
+        return ENOMEM;
+    }
+    uuid_parse(uuid_str, uuid);
+
+    return EOK;
+}
+
+errno_t sec_key_get_uuid(const char *sec_key,
+                         uuid_t uuid)
+{
+    char uuid_str[UUID_STR_SIZE];
+
+    if (!sec_key_valid(sec_key)) {
+        return EINVAL;
+    }
+
+    strncpy(uuid_str, sec_key, UUID_STR_SIZE-1);
+    uuid_str[UUID_STR_SIZE-1] = '\0';
+    uuid_parse(uuid_str, uuid);
+    return EOK;
+}
+
+const char *sec_key_get_name(const char *sec_key)
+{
+    if (!sec_key_valid(sec_key)) {
+        return NULL;
+    }
+
+    return sec_key + UUID_STR_SIZE;
+}
+
+bool sec_key_match_name(const char *sec_key,
+                        const char *name)
+{
+    if (!sec_key_valid(sec_key) || name == NULL) {
+        return false;
+    }
+
+    return strcmp(sec_key + UUID_STR_SIZE, name) == 0;
+}
+
+bool sec_key_match_uuid(const char *sec_key,
+                        uuid_t uuid)
+{
+    errno_t ret;
+    uuid_t key_uuid;
+
+    /* `key_uuid` is output arg and isn't read in sec_key_get_uuid() but
+     * since libuuid is opaque for cppcheck it generates false positive here
+     */
+    /* cppcheck-suppress uninitvar */
```

Comment:
Might just be easier to initialize?  `uuid_t key_uuid = {0};` should work, or you could `memset()`.

Reply:
Depends on the understanding of "easier".

First of all, this is false positive, and addition of any code just for the sake of false positive suppression feels wrong, imo.

Having said this, zero initialization wouldn't hurt, of course. But the problem is ` uuid_t` type is opaque for the user. Using `memset` or other way sets some expectation about the type, which is wrong, imo.

Reply:
> But the problem is `uuid_t` type is opaque for the user.

To split hairs for a moment: `uuid_t` is not opaque to the user or to the compiler; cppcheck just doesn't understand what's going on.  But you can see it in the uuid header - `typedef unsigned char uuid_t[16];` - and that's why initializing it would work.

Reply:
This was a suggestion though - feel free to ignore :)

Reply:
This file contains just old code that was moved out of *json to *key so it can be shared. I moved this change to separate commit.

Reply:
I also added commit to clear the value using uuid_clear(). This did the trick for cppcheck.

Reply:
> > But the problem is `uuid_t` type is opaque for the user.
> 
> To split hairs for a moment: `uuid_t` is not opaque to the user or to the compiler; cppcheck just doesn't understand what's going on. But you can see it in the uuid header - `typedef unsigned char uuid_t[16];` - and that's why initializing it would work.

From my point of view, it is opaque in the sense of clear intent behind library API. The fact that we can access actual implementation doesn't change this. Thus using `uuid_clear()` is better (if it helps).

Another question, if we should do this pessimization at all for the sake of false positive warning.
From my point of view - no, we shouldn't. But since most probably compiler will optimize this away and explanatory comment is still there, that doesn't matter much.

Reply:
You need to do either the comment or the initialization - otherwise your checker output becomes useless.  It seems your preference is to add the comment, which again isn't wrong.

> it is opaque in the sense of clear intent behind library API

If they wanted to hide the type, its implementation doesn't have to be in the header file.  But if you prefer `uuid_clear()` that's again also fine :)